### PR TITLE
TYP: drop

### PR DESCRIPTION
**Pull request by twoertwein**
Tested with mypy and pyright:

```py
import pandas as pd

def bool_() -> bool:
    ...

df = pd.DataFrame({"a": []})
ser = pd.Series({"a": 1, "b": 2})

# DataFrame/Series
reveal_type(df.drop(columns="a"))
reveal_type(df.drop(columns="a", inplace=False))
reveal_type(ser.drop(index="a"))
reveal_type(ser.drop(index="a", inplace=False))

# None
reveal_type(df.drop(columns="a", inplace=True))
reveal_type(ser.drop(index="a", inplace=True))

# both
reveal_type(df.drop(columns="a", inplace=bool_()))
reveal_type(ser.drop(index="a", inplace=bool_()))
```

### CODE REVIEW

File: pandas/_typing.py [hunk 1]
```diff
@@ -297,6 +297,9 @@ def closed(self) -> bool:
 else:
     TakeIndexer = Any
 
+# Shared by functions such as drop and astype
+IgnoreRaise = Literal["ignore", "raise"]
```
Comment (twoertwein) [inline pandas/_typing.py:301]: I also added it to a few other functions (there might be more that only accept ignore/raise).

File: pandas/core/generic.py [hunk 2]
```diff
@@ -4270,16 +4272,59 @@ def reindex_like(
...
+        axis: Axis = ...,
+        index: Hashable | list[Hashable] = ...,
+        columns: Hashable | list[Hashable] = ...,
+        level: Level | None = ...,
+        inplace: bool_t = ...,
+        errors: IgnoreRaise = ...,
+    ) -> NDFrameT | None:
+        ...
+
+    @deprecate_nonkeyword_arguments(version=None, allowed_args=["self", "labels"])
```
Comment (twoertwein) [inline pandas/core/generic.py:4317]: This deprecation is new. Was deprecated in only Series/DataFrame.

File: pandas/plotting/_matplotlib/misc.py [hunk 3]
```diff
@@ -389,7 +389,9 @@ def parallel_coordinates(
             raise ValueError("Length of xticks must match number of columns")
         x = xticks
     else:
-        x = list(range(ncols))
+        # error: Incompatible types in assignment (expression has type "List[int]",
+        # variable has type "Index")
+        x = list(range(ncols))  # type: ignore[assignment]
```
Comment (Dr-Irv) [inline pandas/plotting/_matplotlib/misc.py:394]: Is this needed because now we are using `mypy` 0.941 ?  Doesn't seem related to the rest of the PR.

If so, another possible solution is to declare the type of `x` as `list | Index` before the `if`

File: pandas/core/frame.py [hunk 4]
```diff
@@ -4831,17 +4832,60 @@ def reindex(self, *args, **kwargs) -> DataFrame:
...
+        level: Level | None = ...,
+        inplace: bool = ...,
+        errors: IgnoreRaise = ...,
+    ) -> DataFrame | None:
+        ...
+
+    # error: Signature of "drop" incompatible with supertype "NDFrame"
+    @deprecate_nonkeyword_arguments(version=None, allowed_args=["self", "labels"])
+    def drop(  # type: ignore[override]
+        self,
```
Comment (Dr-Irv) [inline pandas/core/frame.py:4880]: I think we discussed this in another PR, but it's unclear to me why the signatures are incompatible.  Is it because you don't have `self: DataFrame` here?  Or the type of `inplace` is `bool` versus `bool_t` ?  Or something else?

Comment (twoertwein) [reply pandas/core/frame.py:4880]: I think @simonjayhawkins pointed out that this is a result of the parent implementation not having keyword-only arguments, even though the parent overloads use keyword-only arguments. If that is the case, these errors will disappear after the deprecation phase.

`bool` vs `bool_t` has no impact.

File: pandas/plotting/_matplotlib/misc.py (hunk 3 above)
Comment (twoertwein) [reply pandas/plotting/_matplotlib/misc.py:394]: I have no idea why this is needed. But looking at the function, I'm surprised that there wasn't already an ignore: many types are assigned to `x`. I'll try to declare the union before the if-branches.

File: pandas/core/indexes/base.py [hunk 5]
```diff
@@ -6810,7 +6811,7 @@ def insert(self, loc: int, item) -> Index:
         # TODO(2.0) can use Index instead of self._constructor
         return self._constructor._with_infer(new_values, name=self.name)
 
-    def drop(self, labels, errors: str_t = "raise") -> Index:
+    def drop(self, labels, errors: IgnoreRaise = "raise") -> Index:
```
[the remark this answers is not in the document]
Comment (twoertwein) [reply pandas/core/indexes/base.py:6814]: I'll do that. I was just looking for NDFrame/DataFrame/Series.drop and then also adding IgnoreRaise to other functions.  Would definitely make sense to also type Index.drop

File: pandas/core/frame.py (hunk 4 above)
Comment (twoertwein) [reply pandas/core/frame.py:4880]: I was hoping that either the keyword-only issue or the return type (generic vs. the same class) are the cause for that but in this toy example, I do not need an ignore comment - I don't know why. (Probably missing something obvious)

```py
from __future__ import annotations

from typing import (
    Literal,
    TypeVar,
    overload,
)

ParentT = TypeVar("ParentT", bound="Parent")


class Parent:
    @overload
    def test(self, *, a: Literal[False]) -> None:
        ...

    @overload
    def test(self: ParentT, *, a: Literal[True] = ...) -> ParentT:
        ...

    @overload
    def test(self: ParentT, *, a: bool = ...) -> ParentT | None:
        ...

    def test(self: ParentT, a: bool = False) -> ParentT | None:
        ...


class Child(Parent):
    @overload
    def test(self, *, a: Literal[False]) -> None:
        ...

    @overload
    def test(self, *, a: Literal[True] = ...) -> Child:
        ...

    @overload
    def test(self, *, a: bool = ...) -> Child | None:
        ...

    def test(self, a: bool = False) -> Child | None:
        ...
```

File: pandas/plotting/_matplotlib/misc.py (hunk 3 above)
Comment (twoertwein) [reply pandas/plotting/_matplotlib/misc.py]: thanks `list[int] | Index` worked :)

File: pandas/core/indexes/base.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (twoertwein) [reply pandas/core/indexes/base.py]: I typed it based on what the implementation accepts.

File: pandas/core/frame.py (hunk 4 above)
Comment (twoertwein) [reply pandas/core/frame.py:4880]: The decorator seems to irritate mypy, even though the decorator seems to be correctly typed.

Opened https://github.com/python/mypy/issues/12387

Comment (Dr-Irv) [reply pandas/core/frame.py:4880]: Glad your figured it out that it is the decorator.  Can you add a comment in this PR near the `#type: ignore` line that points to the mypy issue?

**Review by Dr-Irv — APPROVED** [review]
Just add a reference to the mypy issue you created so we can track it.

**Review by Dr-Irv — APPROVED** [review]
